Cleaning

idanstark42 opened the pull request:
* Added UT
* Added appveyor.yml
* Added license
* Using new lint
* Removed .npmignore - there is no need for it, .gitignore is enough here
* Fixed typo from middlwere to middleware
* Added more fields to `./lib/log-levels.js`, and made them generate from the basic list of levels.
* Throwing error if trying to set invalid log level
* Using `router` instead of `express` for generic usage
* Removed two unused conditions in `./lib/middleware`

idanstark42 committed:
cleaning

idanstark42 committed:
Started cleaning

idanstark42 committed:
Finished logger UT

idanstark42 committed:
Added body parser

idanstark42 committed:
Finished tests

2roy999 commented on lib/middleware.js:
```diff
@@ -1,35 +1,27 @@
-        break
-      default:
-        req.logger.info(req.body.message)
+    if (LOG_LEVEL_NAMES.includes(level)) {
```
Since when we control the log level with a post request?

idanstark42 replied:
We... dont? This is an endpoint for logging from the client. If you need to write a log from outside the module server, you could send a request here and it wil be logged, without any complications.

2roy999 replied:
ok. I jump to conclusion without checking the function itself. my bad

2roy999 commented on package.json:
```diff
@@ -1,6 +1,7 @@
 {
```
Two things missing:
1. Change the lisence code to `GPL-3.0`
2. Since you removed the npmignore, you should add a files property (This is like npmignore but you are declaring what files you want to be part of the package (it's easier to manage)

idanstark42 replied:
"fields"

idanstark42 replied:
Can we go over which fields should be in the package.json in general, and create a standard for it?

2roy999 replied:
Yes, but not relevant for this PR.

2roy999 commented on package.json:
```diff
@@ -1,6 +1,7 @@
 {
   "name": "@first-lego-league/ms-logger",
   "version": "1.4.1",
+  "description": "Logger for FIRST Lego League System Moduels",
```
Typo: in Modules

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+
+  describe('log', () => {
+    const sandbox = chai.spy.sandbox()
+    const originalDateNow = Date.now
```
Don't mock Date.now. Use `sinon` package `FakeTimers` options.
The idea is that if we decide to change the logger to use moment or even a different Date function we won't need to change this test

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+      const logger = Logger()
+      logger.setLogLevel(LOG_LEVELS.WARN)
+      logger.log(LOG_LEVELS.WARN, 'Some Log')
+      expect(console.log).to.have.been.called.exactly(1)
```
use `once` instead of `exactly(1)`

2roy999 commented on lib/logger.js:
```diff
@@ -19,6 +17,9 @@ function Logger () {
 }
 
 Logger.prototype.setLogLevel = function (logLevel) {
```
If there is a `setLogLevel` there should be a `getLogLevel`.
Even better create getter, setter for logLevel and save the value in `_logLevel`.
For saving compatibility with older why wrap this function with `deprecated` (https://nodejs.org/dist/latest-v10.x/docs/api/util.html#util_util_deprecate_fn_msg_code)

idanstark42 replied:
When saying `getter` and `setter` you're talking about simply functions that set/return the value of the `_logLevel` field, right? Because in that case, the deprecated property is not `setLogLevel` but rather `logLevel`.
In any case, I'm not sure this is in the scope of thie PR. I mean, if I wanted to completly clean this I would also change this into an ES6 class. I didn't mean to change the code any more then I needed to copmlete the expected behaviour of the existing API.

2roy999 replied:
I mean a JS getter/setter : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get.
It might not be in the scope of the PR but we already added a new method `getLogLevel`, so we do add api which we (or a least me) think it's not good. Also, making this an ES6 class also don't take a lot of time so for such small class.

Anyway, I will leave this call for you.

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+    expect(Logger().logLevel).to.equal(LOG_LEVELS.DEBUG)
+  })
+
+  describe('set log level', () => {
```
redundant describe. it's contains to tests regrading a relatively simple process/operation.

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+      const logger = Logger()
+      logger.setLogLevel(LOG_LEVELS.WARN)
+      logger.log(LOG_LEVELS.FATAL, 'Some Log')
+      expect(console.log).to.have.been.called.exactly(1)
```
same as above

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+
+    it('logs the formatter log according the logger\'s value formatting method', () => {
+      const logger = Logger()
+      logger.formatLog = chai.spy(() => { })
```
We don't want to spy on an object own methods. The test shouldn't test if the logger is doing something in a certain way they should test what the logger is doing.

idanstark42 replied:
So how should I test this functionallity?

2roy999 replied:
Take the logged message and check if it's a json object with correct attributes.

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+
+    it('logs the formatter log according the logger\'s value formatting method', () => {
+      const logger = Logger()
+      logger.formatLog = chai.spy(() => { })
```
It's also seems like the function `formatLog` should be protected by and be called `_formatLog`

idanstark42 replied:
I think so too, but again, I didn't want to change the design of the package.

2roy999 replied:
ok

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+    })
+  })
+
+  describe('debug', () => {
```
redundant describe

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+    })
+  })
+
+  describe('info', () => {
```
redundant describe

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+    })
+  })
+
+  describe('warn', () => {
```
redundant describe

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+    })
+  })
+
+  describe('error', () => {
```
redundant describe

2roy999 commented on test/lib/logger.test.js:
```diff
@@ -0,0 +1,169 @@
+    })
+  })
+
+  describe('fatal', () => {
```
redundant describe

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+        .post('/log/debug')
+        .send({ message })
+        .expect(201, () => {
+          expect(mockLogger.log).to.have.been.called.exactly(1)
```
use `once`

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+
+  describe('endpoint', () => {
+    it('only logs a log request once', done => {
+      request(app)
```
`supertest` can be used as a promise. which is much more clear:
```
return request(...)
  .post(...)
  .send(...)
  .expect(201)
  .then(() => expect(mockLogger).to.have.been.called.once)
```

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+    })
+
+    it('logs with the correct level', done => {
+      request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+    })
+
+    it('logs with info if the level does not exist', done => {
+      request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+    it('responds with 500 if the log throws an error', done => {
+      const originalInfo = mockLogger.info
+      mockLogger.info = () => { throw new Error('some error') }
+      request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+    })
+
+    it('only logs the request once', done => {
+      request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+        .post('/some_path')
+        .send({ message })
+        .expect(status, () => {
+          expect(mockLogger.log).to.have.been.called.exactly(1)
```
use once

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+    })
+
+    it('logs in debug level if the status is a successful one', done => {
+      request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+
+    it('logs in warn level if the status is client fault', done => {
+      status = 400
+      request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+
+    it('logs in error level if the status is server fault', done => {
+      status = 500
+      request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+      app.use(middleware)
+      app.use((req, res) => res.sendStatus(status))
+
+      request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+    })
+  })
+
+  describe('middleware', () => {
```
redundant describe

idanstark42 replied:
Why? It even has a beforeEach assigned to it. It's to encapsulate all of the middleware specific tests, as compared to the endpoint (`POST /log/:level`) tests.

2roy999 replied:
A describe block is shouldn't be created to fulfill a `beforeEach` certain tests.
A better approch would be putting this before each  with the before each of all tests, 
that way when one wants to add a new test he have an dummy app which is functionality is full
and can be found at one single place at the beginning of the main describe

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+      })
+
+      it('contains the request method in captial letters when the method is get', done => {
+        request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+      })
+
+      it('contains the request method in captial letters when the method is post', done => {
+        request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+      })
+
+      it('contains the request path', done => {
+        request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+      })
+
+      it('contains the request status code', done => {
+        request(app)
```
use promise

2roy999 commented on test/lib/middleware.test.js:
```diff
@@ -0,0 +1,224 @@
+        })
+    })
+
+    describe('log message', () => {
```
I don't really like a lot of sub-describes. Especially when we check the same function in all tests.
The problem with that is that you move part of the description of the test away from the test itself.
When you read the description of one of the test you need to remember in which of the sub-describe you are now and fill on the blank.

A better approch is to start each test with:
`logs a message contains ...`
which would be read in the tests as:
`Middleware logs a message contains ...'

idanstark42 replied:
But then I would also have to repeat the contents of the `beforeEach` in each of them.

2roy999 replied:
Do this for all tests.
Or even better: why are you doing this and not using a spy?

2roy999 commented on appveyor.yml:
```diff
@@ -0,0 +1,22 @@
+# Post-install test scripts.
+test_script:
+  # Output useful info for debugging.
+  - node --version
```
don't put the node version and the yarn version on the test-script. 
The node version is defined explicitly above and yarn is printing the version in every use.

idanstark42 reviewed (commented):
Fixed some, commented on the rest.

idanstark42 committed:
CR fixes 1

idanstark42 committed:
CR fixes #2

idanstark42 commented:
So i went ahead and changed it to be an ES6 class, with `set` and `get` for the `logLevel`. Also fixed all the other things except the `package.json`, which is waiting, along with the rest of this PR, until we finish the development standards file.

2roy999 commented on lib/logger.js:
```diff
@@ -1,68 +1,64 @@
-  this.setLogLevel(process.env.LOG_LEVEL || LOG_LEVELS.DEBUG)
-}
-
-Logger.prototype.setLogLevel = function (logLevel) {
```
We should either add this method with deprecated properties or make sure next version is major version.

taltaub22 reviewed (approved):
I'm approving this as the changed look fine to me and I agree with JJ comments

idanstark42 committed:
Added files

idanstark42 commented:
I'm sorry for dragging this, but I want this package to be an example for code cleanliness, because a lot of packages rely on it. I'll merge this PR after we do the same cleaning for `ms-correlation` and update this package's dependency of it. But for now, @2roy999 can you check that the `files` property in the `package.json` is correct?

2roy999 commented:
@idanstark42, run `yarn pack` and check if the zip contains the files you want.

idanstark42 commented:
What I meant was if there are files that should be included/excluded and are currently the opposite.

2roy999 commented:
oh,
I think we can remove `package.json`, `README` and `LICENSE`. npm/yarn always include them anyway, so I think it's better if it will focus on the importent files (js).
Also we can remove `yarn.lock` it's no relevence in packages.

idanstark42 committed:
clenaing package.json
